packages/loader/container-loader/src/attachment.ts

```diff
@@ -0,0 +1,211 @@
+	 * @param data - the data to create services from,
+	 * the summary property being the most relevant part of the data.
+	 * @returns A compatible storage service
+	 */
+	readonly createOrGetStorageService: (
+		data: DetachedDataWithOutstandingBlobs | AttachingDataWithBlobs | AttachingDataWithoutBlobs,
+	) => Promise<Pick<IDocumentStorageService, "createBlob" | "uploadSummaryWithContext">>;
+
+	/**
+	 * The detached blob storage if it exists.
+	 */
+	readonly detachedBlobStorage?: Pick<IDetachedBlobStorage, "getBlobIds" | "readBlob" | "size">;
+
+	/**
+	 * The caller should create the attachment summary for the container.
+	 * @param redirectTable - Maps local blob ids to remote blobs ids.
+	 * @returns The attachment summary for the container.
+	 */
+	readonly createAttachmentSummary: (
+		redirectTable?: Map<string, string>,
+	) => CombinedAppAndProtocolSummary;
+
+	/**
+	 * Whether offline load is enabled or not.
+	 */
+	readonly offlineLoadEnabled: boolean;
+}
+
+/**
+ * Executes the attach process state machine based on the provided data and services.
+ * This method is retriable on failure. Based on the provided initialAttachmentData
+ * this method will resume the attachment process and attempt to complete it.
+ *
+ * @param props - The data and services necessary to run the attachment process
+ */
+export const runRetriableAttachProcess = async (props: AttachProcessProps): Promise<void> => {
+	let currentData: AttachmentData = props.initialAttachmentData;
+
+	if (currentData.blobs === undefined) {
+		// If attachment blobs were uploaded in detached state we will go through a different attach flow
+		const outstandingAttachmentBlobs =
+			props.detachedBlobStorage !== undefined && props.detachedBlobStorage.size > 0;
+		// Determine the next phase of attaching which depends on if there are attachment blobs
+		// if there are, we will stay detached, so an empty file can be created, and the blobs
+		// uploaded, otherwise we will get the summary to create the file with and move to attaching
+		currentData = outstandingAttachmentBlobs
+			? {
+					state: AttachState.Detached,
+					blobs: "outstanding",
+					redirectTable: new Map<string, string>(),
+			  }
+			: {
+					state: AttachState.Attaching,
+					summary: props.createAttachmentSummary(),
+					blobs: "none",
+			  };
+		props.setAttachmentData(currentData);
+	}
+
+	// this has to run here, as it is what creates the file
+	// and we need to file for all possible cases after this point
+	const storage = await props.createOrGetStorageService(currentData);
+
+	if (currentData.blobs === "outstanding") {
+		const { redirectTable } = currentData;
+		// upload blobs to storage
+		assert(!!props.detachedBlobStorage, 0x24e /* "assertion for type narrowing" */);
+
+		// build a table mapping IDs assigned locally to IDs assigned by storage and pass it to runtime to
+		// support blob handles that only know about the local IDs
+		while (redirectTable.size < props.detachedBlobStorage.size) {
```

Comment:
IDetachedBlobStorage is provided to the loaded by the developer using fluid, so we can't add logic to it. maybe we could encapsulate in the storage adapter, i need to see if that would be easy, otherwise may just file a follow up